internal/engine/exit/controller.go

```diff
@@ -2,124 +2,331 @@ package exit
+				allResourcesOK = false
+			}
+		case tiltruns.RuntimeTypeJob:
+			if res.Runtime.Status != tiltruns.RuntimeStatusSucceeded {
+				allResourcesOK = false
+			}
+		default:
+			status.Done = true
+			status.Error = fmt.Sprintf("unsupported resource type: %s", res.Runtime.Type)
+			return
 		}
 	}
 
-	// If all the resources are OK, we're done.
-	if len(state.ManifestTargets) > 0 &&
-		state.InitialBuildsCompleted() && allOK {
-		return Action{ExitSignal: true}
+	if allResourcesOK && len(status.Resources) != 0 {
+		status.Done = true
 	}
+}
 
-	return Action{}
+func (c *Controller) updateStatus(ctx context.Context, state store.EngineState) error {
+	// TODO(milas): Tiltfile is not included yet
+	newStatus := &tiltruns.TiltRunStatus{
+		PID:       c.pid,
+		StartTime: metav1.NewMicroTime(c.startTime),
+		Tiltfile:  apiTiltfileStateFromEngine(state.TiltfileState),
+		Resources: make([]tiltruns.ResourceState, 0, len(state.ManifestTargets)),
+	}
+
+	for _, mt := range state.ManifestTargets {
+		newStatus.Resources = append(newStatus.Resources, apiManifestStateFromEngine(mt))
+	}
+
+	processExitCondition(c.tiltRun.Spec.ExitCondition, newStatus)
+
+	if equality.Semantic.DeepEqual(c.tiltRun.Status, newStatus) {
+		return nil
+	}
+
+	updated := c.tiltRun.DeepCopy()
+	newStatus.DeepCopyInto(&updated.Status)
+
+	if err := c.client.Status().Update(ctx, updated); err != nil {
+		return err
+		//return fmt.Errorf("failed to update TiltRun status: %v", err)
+	}
+
+	c.tiltRun = updated
+
+	return nil
 }
 
-func (c *Controller) isRuntimeDone(mt *store.ManifestTarget) bool {
-	rs := mt.State.RuntimeState
-	if rs == nil {
-		return false
-	}
-
-	status := rs.RuntimeStatus()
-	statusOK := status == model.RuntimeStatusOK || status == model.RuntimeStatusNotApplicable
-	if !statusOK {
-		return false
-	}
-
-	// If this is a job, check to see if it has run to completion
-	//
-	// TODO(nick): This is...not great. Ideally, Tilt would track the status of
-	// every resource type it deploys, then we'd have some general-purpose system
-	// for expressing success criteria on different resource types (like
-	// https://www.openpolicyagent.org/). This is just a hack to make this work
-	// for jobs, until it makes sense to build out that type-independent
-	// infrastructure.
-	isK8s := mt.Manifest.IsK8s()
-	isK8sJob := isK8s && mt.Manifest.K8sTarget().HasJob()
-	if isK8sJob {
-		k8sState, ok := mt.State.RuntimeState.(store.K8sRuntimeState)
-		if !ok {
-			return false
+func (c *Controller) shouldExit() Action {
+	if c.tiltRun.Status.Done {
+		var exitErr error
+		if c.tiltRun.Status.Error != "" {
+			exitErr = errors.New(c.tiltRun.Status.Error)
 		}
+		return Action{ExitSignal: true, ExitError: exitErr}
+	}
 
-		pod := k8sState.MostRecentPod()
-		if pod.Phase != v1.PodSucceeded {
-			return false
+	return Action{}
+}
+
+func (c *Controller) OnChange(ctx context.Context, st store.RStore, _ store.ChangeSummary) {
+	state := st.RLockState()
```

Comment:
i'm a little bit nervous about holding the EngineState RLock for this entire function, which will prevent any writes while it's running

Reply:
Refactored in latest revision to hold the lock for much shorter duration